In [8]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

2. READ THE DATA INTO A DATAFRAME, AND GET THE SHAPE OF THE DATA AND THE FIRST 5 RECORDS.

In [9]:
#read the data

df = pd.read_csv('news.csv')

#get shape and head

df.shape
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

3. GET THE LABELS FROM THE DATAFRAME

In [11]:
#Get the labels

labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

4. SPLIT THE DATASET INTO TRAINING AND TESTING SETS

In [28]:
#Split the dataset

x_train, x_test, y_train, y_test = train_test_split(df['text'], labels, test_size=0.2, random_state=7)

5. INITIALIZE A TfidfVectorizer WITH STOP WORDS. TfidfVectorizer TURNS A COLLECTION OF RAW DOCUMENTS INTO A MATRIX OF TF-IDF FEATURES.

In [29]:
#Initialize a TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer(stop_words = 'turkish', max_df=0.7)

#Fit and transform train set, transform test set
tfidf_train = tfidf_vectorizer.fit_transform(x_train)
tfidf_test = tfidf_vectorizer.transform(x_test)

6. INITIALIZE A PASSIVEAGGRESSIVECLASSIFIER, FIT THIS ON tfidf_train and y_train

In [30]:
#initialize a PassiveAggressiveClassifier

pclassifier = PassiveAggressiveClassifier(max_iter=50)
pclassifier.fit(tfidf_train, y_train)

#predict on the test set and calculate accuracy

y_predict = pclassifier.predict(tfidf_test)
score = accuracy_score(y_test, y_predict)
print(f'Accurancy: {round(score*100,2)}%')

Accurancy: 92.9%


7. PRINT OUT A CONFUSION MATRIX TO GAIN INSIGHT INTO THE NUMBER OF FALSE AND TRUE NEGATIVES AND POSITIVES.

In [31]:
#build confusion matrix

confusion_matrix(y_test, y_predict, labels =['FAKE', 'REAL'])

array([[589,  49],
       [ 41, 588]], dtype=int64)

with this model, there are 589 true positives, 588 true negatives, 41 false positives, 49 false negatives